### PR TITLE
Convenience Store Payment

### PR DESCRIPTION
**Pull request by jacstn**
#### 1. Objective
This PR enables Convenience Store Payment.

**Note** Payment is available only for customers with Japanese account.
More info about Convenience Store Payment is available here: https://www.omise.co/conbini-pay-easy-online-banking

#### 2. Description of change

Added Virtual and PHP classes to support Convenience Store Payment.
Added `javascript` and `HTML` UI rendering files.
Added new UI Observer which gather user's telephone number and email address.

Checkout page:
image

Admin Panel:
image

Checkout Success page
image

**Note**
Tuning of checkout success page will be done in separated PR

**Note**
There are special conditions of email format that are not developed in this PR.
Convenience store does not support with some valid email addresses.
Magento 2 UI does not filter those email out yet, it will be covered in future PR.

ie. valid email address: tom+243432@gmail.com will accepted by HTML form, but payment will not be successful, and misleading message will be displayed:
image

Following validation is currently supported:
- check if an entered phone number is digits only, and 10-11 characters length,
- if an email address is valid according to general RFC specification,

**Note**
Sending confirmation email with a link to payment instructions will be done in seperated PR.

#### 3. Quality assurance

**🔧 Environments:**

- **Platform version**: Magento CE 2.2.5.
- **Omise plugin version**: Omise-Magento 2.2.42
- **PHP version**: 7.0.29.

**✏️ Details:**
_You need to use keys from Japanese Omise account._

Turn on a `Convenience Store Payment` in the admin panel, check if payment is available on the checkout page.

Turn off a `Convenience Store Payment` method on the admin panel, the option should not be listed on the checkout page anymore.

Make an order and successful payment with the `Convenience Store Payment` method.

Verify if the proper amount and payment method are displayed in [Omise Dashboard](https://dashboard.omise.co)

image

#### 4. Impact of the change

New Payment Method is available for Japanese customers using Magento 2.

#### 5. Priority of change

Normal

#### 6. Additional Notes

none

### CODE REVIEW

File: Gateway/Request/APMBuilder.php [hunk 1]
```diff
@@ -9,13 +9,14 @@
 use Omise\Payment\Model\Config\Internetbanking;
 use Omise\Payment\Model\Config\Tesco;
 use Omise\Payment\Model\Config\Truemoney;
+use Omise\Payment\Model\Config\Conviniencestore;
```
Comment (jonrandy) [inline Gateway/Request/APMBuilder.php]: Convenience is misspelt throughout the code

File: Gateway/Request/APMBuilder.php [hunk 2]
```diff
@@ -9,13 +9,14 @@
...
 use Omise\Payment\Observer\InstallmentDataAssignObserver;
 use Omise\Payment\Observer\InternetbankingDataAssignObserver;
 use Omise\Payment\Observer\TruemoneyDataAssignObserver;
+use Omise\Payment\Observer\ConviniencestoreDataAssignObserver;
 
 class APMBuilder implements BuilderInterface
 {
-
     /**
      * @var string
```
Comment (jonrandy) [inline Gateway/Request/APMBuilder.php]: Should be @const ?

File: etc/adminhtml/system.xml [hunk 3]
```diff
@@ -140,6 +139,21 @@
                         <config_path>payment/omise_offsite_truemoney/title</config_path>
                     </field>
                 </group>
+                <group id="omise_offsite_conveniencestore" translate="label" sortOrder="290" showInDefault="1" showInWebsite="1" showInStore="1">
+                    <label>Enable Convenience Store</label>
```
Comment (jonrandy) [inline etc/adminhtml/system.xml]: Enable Convenience Store Payment

File: Gateway/Request/APMBuilder.php (hunk 2 above)
Comment (jacstn) [reply Gateway/Request/APMBuilder.php]: hmm.. never saw, I think proper way is putting `var` in doc

File: etc/config.xml [hunk 4]
```diff
@@ -74,6 +74,18 @@ xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Store:etc/config.xsd">
                 <payment_action>authorize_capture</payment_action>
             </omise_offsite_truemoney>
 
+            <omise_offsite_conveniencestore>
+                <active>0</active>
+                <title>Convenience Store</title>
+                <model>OmiseConveniencestoreAdapter</model>
```
Comment (jonrandy) [inline etc/config.xml]: Casing seems off in the model name, but I guess it's consistent with what you've done already. `OmiseConvenienceStoreAdapter` would probably have been better

Comment (jacstn) [reply etc/config.xml]: followed existing naming.
`<model>OmiseInternetbankingAdapter</model>`

Comment (jacstn) [reply etc/config.xml]: but that is something we could refactor at some point.

**Comment by jacstn** [comment]
ready to approve? @guzzilar @jonrandy ?

**Comment by jacstn** [comment]
ready to approve @jonrandy @guzzilar ?

**Comment by guzzilar** [comment]
@jacstn I was waiting for the js-refactoring thing 😂  
Or is this ok to be reviewed & merged first?

**Comment by jacstn** [comment]
yes.. let's review and merge it first..

**Comment by jacstn** [comment]
@jonrandy @guzzilar , Ready to approve?

File: Gateway/Request/APMBuilder.php (hunk 2 above)
Comment (jonrandy) [reply Gateway/Request/APMBuilder.php]: Magento 2 docs seem to suggest putting neither. Just a comment if necessary

File: view/frontend/layout/checkout_onepage_success.xml [hunk 5]
```diff
@@ -2,8 +2,10 @@
 <page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
     <body>
         <referenceContainer name="order.success.additional.info">
-            <block class="Omise\Payment\Block\Checkout\Onepage\Success\TescoAdditionalInformation" name="onepage.success.billing_agreement" 
-            template="Omise_Payment::checkout/onepage/success/tesco_additional_info.phtml"/>
+            <block class="Omise\Payment\Block\Checkout\Onepage\Success\TescoAdditionalInformation" name="onepage.success.tesco"
```
Comment (guzzilar) [inline view/frontend/layout/checkout_onepage_success.xml]: Just wonder if renaming one to another won't affect anything to the previous version of plugin?
Does this require merchant to clear store's caches or something?

File: Gateway/Request/APMBuilder.php [hunk 6]
```diff
@@ -9,13 +9,14 @@
 use Omise\Payment\Model\Config\Internetbanking;
 use Omise\Payment\Model\Config\Tesco;
 use Omise\Payment\Model\Config\Truemoney;
+use Omise\Payment\Model\Config\Conveniencestore;
 use Omise\Payment\Observer\InstallmentDataAssignObserver;
 use Omise\Payment\Observer\InternetbankingDataAssignObserver;
 use Omise\Payment\Observer\TruemoneyDataAssignObserver;
+use Omise\Payment\Observer\ConveniencestoreDataAssignObserver;
```
Comment (guzzilar) [inline Gateway/Request/APMBuilder.php]: Just very minor but I think you want to keep it in alphabetically order (guessing from the current order)

File: Gateway/Response/PaymentDetailsHandler.php [hunk 7]
```diff
@@ -1,9 +1,7 @@
 <?php
 namespace Omise\Payment\Gateway\Response;
-
```
Comment (guzzilar) [inline Gateway/Response/PaymentDetailsHandler.php]: Maybe this was accidentally formatted by a text-editor?

File: view/frontend/web/js/view/payment/method-renderer/omise-offline-conveniencestore-method.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+             * Get total amount of an order
+             *
+             * @return {integer}
+             */
+            getTotal: function () {
+                return + window.checkoutConfig.totalsData.grand_total;
+            },
+
+            /**
+             * Get customer phone number saved in profile
```
Comment (guzzilar) [inline view/frontend/web/js/view/payment/method-renderer/omise-offline-conveniencestore-method.js]: Get customer name?

File: view/frontend/web/js/view/payment/method-renderer/omise-offline-conveniencestore-method.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        fullScreenLoader,
+        redirectOnSuccessAction,
+        quote,
+        customer,
+        errorProcessor,
+        priceUtils
+    ) {
+        'use strict';
+
+        const convStoreMinimumPurchaseAmount = 150;
```
Comment (guzzilar) [inline view/frontend/web/js/view/payment/method-renderer/omise-offline-conveniencestore-method.js]: minimum amount has recently been changed to `200` from the API side.

File: view/frontend/web/css/styles.css [hunk 10]
```diff
@@ -100,3 +100,12 @@
...
     margin-bottom:18px;
 }
+
+#omise_offline_conveniencestorephoneNumber {
+    width: 225px;
+}
+
+#omise_offline_conveniencestoreemail {
+    width: 225px;
+    margin-bottom:18px;
```
Comment (guzzilar) [inline view/frontend/web/css/styles.css]: As Convenience Store payment and TrueMoney Wallet are sharing the similar stylesheet. Maybe we can make a generic class for this?

File: view/frontend/web/css/styles.css [hunk 11]
```diff
@@ -100,3 +100,12 @@
...
+}
```
Comment (guzzilar) [inline view/frontend/web/css/styles.css]: One breakline here?

File: Gateway/Request/APMBuilder.php (hunk 6 above)
Comment (jacstn) [reply Gateway/Request/APMBuilder.php]: nooo... it was mixed between the time of adding, and randomness, internet banking/true money/convenience store/ that was the time of adding.

File: view/frontend/layout/checkout_onepage_success.xml (hunk 5 above)
Comment (jacstn) [reply view/frontend/layout/checkout_onepage_success.xml]: no threat. 
it was mistake in the past, and it was the same name as paypal.
That was probably bringing problems.
We need to change it, but we could fix it in separated PR if that is what would be better here. @guzzilar

File: Gateway/Request/APMBuilder.php (hunk 6 above)
Comment (guzzilar) [reply Gateway/Request/APMBuilder.php]: I thought we re-ordered it to be in alphabetically order when we introduced this APMBuilder class 😂  (as Alipay comes first, then follows by Installment and InternetBanking).
However, it's just super minor. I'll mark it as resolve : )

File: view/frontend/layout/checkout_onepage_success.xml (hunk 5 above)
Comment (guzzilar) [reply view/frontend/layout/checkout_onepage_success.xml]: 🙆‍♂

**Comment by jacstn** [comment]
I think all done @guzzilar

File: view/frontend/web/js/view/payment/method-renderer/omise-offline-conveniencestore-method.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+                return + window.checkoutConfig.totalsData.grand_total;
+            },
+
+            /**
+             * Get customer name saved in profile
+             *
+             * @return {string}
+             */
+            getConvenienceStoreCustomersName: function() {
+                return customer().customer().fullname;
```
[the remark this answers is not in the document]
Comment (jacstn) [reply view/frontend/web/js/view/payment/method-renderer/omise-offline-conveniencestore-method.js]: I believe it is fixed now, @guzzilar

File: view/frontend/web/template/payment/offline-conveniencestore-form.html [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    <div class="payment-method-title field choice">
+        <input type="radio" name="payment[method]" class="radio"
+            data-bind="attr: {
+               'id': getCode()
+               },
+               value: getCode(),
+               checked: isChecked,
+               click: selectPaymentMethod,
+               visible: isRadioButtonVisible(),
+               enable: isActive()" />
```
[the remark this answers is not in the document]
Comment (jacstn) [reply view/frontend/web/template/payment/offline-conveniencestore-form.html]: 👍

**Comment by jacstn** [comment]
yes.. that will be done in the next PR due to JS refactoring.

**Comment by jacstn** [comment]
thank you @guzzilar for your all inputs in this PR 👍

**Pull request merged**